test/Constraints/casts_swift6.swift

```diff
@@ -27,9 +27,7 @@ func test_compatibility_coercions(_ arr: [Int], _ optArr: [Int]?, _ dict: [Strin
    // expected-note@-1 {{arguments to generic parameter 'Element' ('Int' and 'String') are expected to be equal}}
 
   // Make sure we error on the following in Swift 6 mode.
-
-  // FIXME: Bad diagnostics (SR-15843)
-  _ = id(arr) as [String] // expected-error {{type of expression is ambiguous without more context}}
+  _ = id(arr) as [String] // expected-error {{conflicting arguments to generic parameter 'T' ('[Int]' vs. '[String]')}}
   _ = (arr ?? []) as [String] // expected-error {{type of expression is ambiguous without more context}}
   _ = (arr ?? [] ?? []) as [String] // expected-error {{type of expression is ambiguous without more context}}
   _ = (optArr ?? []) as [String] // expected-error {{type of expression is ambiguous without more context}}
```

Comment:
Thanks for fixing this! I was kind of hoping that the fix for the `id(arr) as [String]` would also fix these operator cases too. It seems they also produce a `GenericArgumentsMismatch` fix, so I'm curious why `diagnoseConflictingGenericArguments` doesn't also now handle them? I'm happy if you want to look into that as a follow-up tho.

Reply:
Ah I see the logic in `diagnoseConflictingGenericArguments` is iterating over `cs.getTypeVariables()`, so seems like it might miss type variables from solutions if the constraint system was off exploring a scope with different type variables. Shouldn't we be collecting type variables straight from the solutions there @xedin?

Reply:
It's likely because the condition only accepts solution if it has only the fix kinds listed in the `return`, and I'm pretty sure operators also produce conformance requirement fixes for their generic overloads, we might want to relax the condition here a bit to diagnose them...

Reply:
Interesting, regardless of that issue we need to switch the logic from `cs.getTypeVariables()` to `solution.typeBindings`

Reply:
Sorry for that, I think I should have taken a deeper look at the other cases... thanks for the comments @hamishknight @xedin I have made the adjustments :)